### PR TITLE
Fixes for rockstar under python3. closes #1412

### LINKED ISSUE #1412: Rockstar interface broken with python 3
### Bug report

**Bug summary**

Running the nested grid cookbook recipe for finding halos with rockstar fails.
Works (no abort) using python 2.

**Code for reproduction**

see http://yt-project.org/docs/dev/cookbook/rockstar_nest.py linked here http://yt-project.org/docs/dev/cookbook/cosmological_analysis.html#running-rockstar-to-find-halos-on-multi-resolution-particle-datasets

**Actual outcome**

```
yt : [INFO     ] 2017-05-20 11:40:57,264 Global parallel computation enabled: 2 / 3
yt : [INFO     ] 2017-05-20 11:40:57,264 Global parallel computation enabled: 0 / 3
yt : [INFO     ] 2017-05-20 11:40:57,264 Global parallel computation enabled: 1 / 3
P000 yt : [INFO     ] 2017-05-20 11:40:58,783 Parameters: current_time              = 14.1336338797
P000 yt : [INFO     ] 2017-05-20 11:40:58,783 Parameters: domain_dimensions         = [128 128 128]
P000 yt : [INFO     ] 2017-05-20 11:40:58,793 Parameters: domain_left_edge          = [ 0.  0.  0.]
P000 yt : [INFO     ] 2017-05-20 11:40:58,794 Parameters: domain_right_edge         = [ 1.  1.  1.]
P000 yt : [INFO     ] 2017-05-20 11:40:58,794 Parameters: cosmological_simulation   = 1
P000 yt : [INFO     ] 2017-05-20 11:40:58,795 Parameters: current_redshift          = 14.092558914923
P000 yt : [INFO     ] 2017-05-20 11:40:58,795 Parameters: omega_lambda              = 0.6911
P000 yt : [INFO     ] 2017-05-20 11:40:58,795 Parameters: omega_matter              = 0.3089
P000 yt : [INFO     ] 2017-05-20 11:40:58,795 Parameters: hubble_constant           = 0.6774
P000 yt : [INFO     ] 2017-05-20 11:40:58,861 Starting 'Parsing Hierarchy '
P002 yt : [INFO     ] 2017-05-20 11:40:58,861 Starting 'Parsing Hierarchy '
P001 yt : [INFO     ] 2017-05-20 11:40:58,861 Starting 'Parsing Hierarchy '
P002 yt : [INFO     ] 2017-05-20 11:40:58,956 Finishing 'Parsing Hierarchy '
P001 yt : [INFO     ] 2017-05-20 11:40:58,958 Finishing 'Parsing Hierarchy '
P000 yt : [INFO     ] 2017-05-20 11:40:58,961 Finishing 'Parsing Hierarchy '
P000 yt : [INFO     ] 2017-05-20 11:40:59,146 Gathering a field list (this may take a moment.)
[hpost3:191042] 2 more processes have sent help message help-mpi-btl-openib.txt / default subnet prefix
[hpost3:191042] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
P000 yt : [INFO     ] 2017-05-20 11:41:08,858 The citation for the Rockstar halo finder can be found at
P000 yt : [INFO     ] 2017-05-20 11:41:08,858 http://adsabs.harvard.edu/abs/2013ApJ...762..109B
P001 yt : [INFO     ] 2017-05-20 11:41:08,858 Rockstar is using 1 readers and 1 writers
P002 yt : [INFO     ] 2017-05-20 11:41:08,858 Rockstar is using 1 readers and 1 writers
P000 yt : [INFO     ] 2017-05-20 11:41:08,858 Rockstar is using 1 readers and 1 writers
P001 yt : [WARNING  ] 2017-05-20 11:42:19,993 Total Particle Count: 4.533e+05
  File "rockstar_nest.py", line 59, in <module>
    hc.create()
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_catalog.py", line 335, in create
    self._run(save_halos, save_catalog, njobs=njobs, dynamic=dynamic)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/utilities/parallel_tools/parallel_analysis_interface.py", line 306, in barrierize
    retval = func(*args, **kwargs)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_catalog.py", line 404, in _run
    self.halos_ds = self.finder_method(self.data_ds)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_finding_methods.py", line 44, in __call__
    return self.function(ds, *self.args, **self.kwargs)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_finding_methods.py", line 77, in _rockstar_method
    rh.run()
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_finding/rockstar/rockstar.py", line 341, in run
    min_halo_size = self.min_halo_size)
  File "yt/analysis_modules/halo_finding/rockstar/rockstar_interface.pyx", line 249, in yt.analysis_modules.halo_finding.rockstar.rockstar_interface.RockstarInterface.setup_rockstar (yt/analysis_modules/halo_finding/rockstar/rockstar_interface.c:4291)
P001 yt : [ERROR    ] 2017-05-20 11:42:20,193 TypeError: expected bytes, str found
  File "rockstar_nest.py", line 59, in <module>
    hc.create()
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_catalog.py", line 335, in create
    self._run(save_halos, save_catalog, njobs=njobs, dynamic=dynamic)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/utilities/parallel_tools/parallel_analysis_interface.py", line 306, in barrierize
    retval = func(*args, **kwargs)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_catalog.py", line 404, in _run
    self.halos_ds = self.finder_method(self.data_ds)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_finding_methods.py", line 44, in __call__
    return self.function(ds, *self.args, **self.kwargs)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_finding_methods.py", line 77, in _rockstar_method
    rh.run()
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_finding/rockstar/rockstar.py", line 341, in run
    min_halo_size = self.min_halo_size)
  File "yt/analysis_modules/halo_finding/rockstar/rockstar_interface.pyx", line 249, in yt.analysis_modules.halo_finding.rockstar.rockstar_interface.RockstarInterface.setup_rockstar (yt/analysis_modules/halo_finding/rockstar/rockstar_interface.c:4291)
P000 yt : [ERROR    ] 2017-05-20 11:42:20,193 TypeError: expected bytes, str found
  File "rockstar_nest.py", line 59, in <module>
    hc.create()
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_catalog.py", line 335, in create
    self._run(save_halos, save_catalog, njobs=njobs, dynamic=dynamic)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/utilities/parallel_tools/parallel_analysis_interface.py", line 306, in barrierize
    retval = func(*args, **kwargs)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_catalog.py", line 404, in _run
    self.halos_ds = self.finder_method(self.data_ds)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_finding_methods.py", line 44, in __call__
    return self.function(ds, *self.args, **self.kwargs)
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_analysis/halo_finding_methods.py", line 77, in _rockstar_method
    rh.run()
  File "/gfs2/work/nipgrete/yt-conda/src/yt-git/yt/analysis_modules/halo_finding/rockstar/rockstar.py", line 341, in run
    min_halo_size = self.min_halo_size)
  File "yt/analysis_modules/halo_finding/rockstar/rockstar_interface.pyx", line 249, in yt.analysis_modules.halo_finding.rockstar.rockstar_interface.RockstarInterface.setup_rockstar (yt/analysis_modules/halo_finding/rockstar/rockstar_interface.c:4291)
P002 yt : [ERROR    ] 2017-05-20 11:42:20,193 TypeError: expected bytes, str found
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 1 in communicator MPI_COMM_WORLD 
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------

```

**Expected outcome**

No error.

**Version Information**
  * Operating System: SUSE Linux Enterprise Server 11 (x86_64)
  * Python Version: 3 
  * yt version: c9a645d

dependencies with conda, yt from source

### PR DESCRIPTION
**Pull request by ngoldbaum**
@pgrete it would be great if you could test this.

### CODE REVIEW

**Comment by pgrete** [comment]
I get a different error now:
"""
yt : [INFO     ] 2017-05-26 22:00:07,278 Global parallel computation enabled: 2 / 3
yt : [INFO     ] 2017-05-26 22:00:07,279 Global parallel computation enabled: 0 / 3
yt : [INFO     ] 2017-05-26 22:00:07,279 Global parallel computation enabled: 1 / 3
P000 yt : [INFO     ] 2017-05-26 22:00:07,363 Parameters: current_time              = 14.1336338797
P000 yt : [INFO     ] 2017-05-26 22:00:07,364 Parameters: domain_dimensions         = [128 128 128]
P000 yt : [INFO     ] 2017-05-26 22:00:07,364 Parameters: domain_left_edge          = [ 0.  0.  0.]
P000 yt : [INFO     ] 2017-05-26 22:00:07,365 Parameters: domain_right_edge         = [ 1.  1.  1.]
P000 yt : [INFO     ] 2017-05-26 22:00:07,366 Parameters: cosmological_simulation   = 1
P000 yt : [INFO     ] 2017-05-26 22:00:07,366 Parameters: current_redshift          = 14.092558914923
P000 yt : [INFO     ] 2017-05-26 22:00:07,366 Parameters: omega_lambda              = 0.6911
P000 yt : [INFO     ] 2017-05-26 22:00:07,366 Parameters: omega_matter              = 0.3089
P000 yt : [INFO     ] 2017-05-26 22:00:07,366 Parameters: hubble_constant           = 0.6774
P001 yt : [INFO     ] 2017-05-26 22:00:07,370 Starting 'Parsing Hierarchy '
P000 yt : [INFO     ] 2017-05-26 22:00:07,370 Starting 'Parsing Hierarchy '
P002 yt : [INFO     ] 2017-05-26 22:00:07,379 Starting 'Parsing Hierarchy '
P000 yt : [INFO     ] 2017-05-26 22:00:07,471 Finishing 'Parsing Hierarchy '
P001 yt : [INFO     ] 2017-05-26 22:00:07,498 Finishing 'Parsing Hierarchy '
P002 yt : [INFO     ] 2017-05-26 22:00:07,531 Finishing 'Parsing Hierarchy '
P000 yt : [INFO     ] 2017-05-26 22:00:07,710 Gathering a field list (this may take a moment.)
[hpost3:25947] 2 more processes have sent help message help-mpi-btl-openib.txt / default subnet prefix
[hpost3:25947] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
P000 yt : [INFO     ] 2017-05-26 22:00:16,092 The citation for the Rockstar halo finder can be found at
P002 yt : [INFO     ] 2017-05-26 22:00:16,092 Rockstar is using 1 readers and 1 writers
P001 yt : [INFO     ] 2017-05-26 22:00:16,092 Rockstar is using 1 readers and 1 writers
P000 yt : [INFO     ] 2017-05-26 22:00:16,092 http://adsabs.harvard.edu/abs/2013ApJ...762..109B
P000 yt : [INFO     ] 2017-05-26 22:00:16,092 Rockstar is using 1 readers and 1 writers
P001 yt : [WARNING  ] 2017-05-26 22:01:21,146 Total Particle Count: 4.533e+05
[Error] Couldn't open :!  (Err: Servname not supported for ai_socktype)
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[2154,1],0]
  Exit code:    1
--------------------------------------------------------------------------
"""

**Comment by ngoldbaum** [comment]
@pgrete unfortunately I know basically next to nothing about rockstar. Do you have any idea what might be the way forward here?

**Comment by pgrete** [comment]
I'm also not really familiar with the machinery. Getting sensible results soon than later has priority at this point so I might stick to the python 2 interface for now.

File: yt/analysis_modules/halo_finding/rockstar/rockstar.py [hunk 1]
```diff
@@ -325,7 +325,9 @@ def run(self, block_ratio = 1, callbacks = None, restart = False):
             self.ts._pre_outputs = self.ts._pre_outputs[restart_num:]
         else:
             restart_num = 0
-        self.handler.setup_rockstar(self.server_address, self.port,
+        self.handler.setup_rockstar(
+                    self.server_address.encode('ascii'),
```
Comment (Xarthisius) [inline yt/analysis_modules/halo_finding/rockstar/rockstar.py]: I think if you use `six.b(self.server_address)` it's going to work for both py2 and py3. Similarly for all the other strings

**Comment by ngoldbaum** [comment]
@Xarthisius thanks so much! That appears to have fixed the tests. Perhaps we should try to figure out why the rockstar tests weren't running on python3 before? Or perhaps we should merge this now and come back to that.

**Pull request merged**